xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -1113,6 +1138,270 @@ public String toString() {
+    }
+    if (rawFaultFilterConfig.getTypeUrl().equals(
+        "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault")) {
+      HTTPFault httpFaultProto;
+      try {
+        httpFaultProto = rawFaultFilterConfig.unpack(HTTPFault.class);
+      } catch (InvalidProtocolBufferException e) {
+        return StructOrError.fromError("Invalid proto: " + e);
+      }
+      return HttpFault.fromEnvoyProtoHttpFault(httpFaultProto);
+    }
+    return StructOrError.fromError(
+        "Invalid type url for envoy.fault filter: " + rawFaultFilterConfig.getTypeUrl());
+  }
+
+  static boolean parseHttpFaultFilter() {
+    return enableFaultInjection;
+  }
+
+  /**
+   * See corresponding Envoy proto message {@link
+   * io.envoyproxy.envoy.extensions.filters.http.fault.v3.HTTPFault}.
+   */
+  static final class HttpFault {
+    @Nullable
+    final FaultDelay faultDelay;
+    @Nullable
+    final FaultAbort faultAbort;
+
+    HttpFault(@Nullable FaultDelay faultDelay, @Nullable FaultAbort faultAbort) {
+      this.faultDelay = faultDelay;
+      this.faultAbort = faultAbort;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      HttpFault httpFault = (HttpFault) o;
+      return Objects.equals(faultDelay, httpFault.faultDelay)
+          && Objects.equals(faultAbort, httpFault.faultAbort);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(faultDelay, faultAbort);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("faultDelay", faultDelay)
+          .add("faultAbort", faultAbort)
+          .toString();
+    }
+
+    static StructOrError<HttpFault> fromEnvoyProtoHttpFault(HTTPFault httpFault) {
+      FaultDelay faultDelay = null;
+      FaultAbort faultAbort = null;
+      if (httpFault.hasDelay()) {
+        faultDelay = FaultDelay.fromEnvoyProtoFaultDelay(httpFault.getDelay());
+      }
+      if (httpFault.hasAbort()) {
+        StructOrError<FaultAbort> faultAbortOrError =
+            FaultAbort.fromEnvoyProtoFaultAbort(httpFault.getAbort());
+        if (faultAbortOrError.getErrorDetail() != null) {
+          return StructOrError.fromError(
+              "HttpFault contains invalid FaultAbort: " + faultAbortOrError.getErrorDetail());
+        }
+        faultAbort = faultAbortOrError.getStruct();
+      }
+      if (faultDelay == null && faultAbort == null) {
+        return StructOrError.fromError(
+            "Invalid HttpFault: neither fault_delay nor fault_abort is specified");
+      }
+      return StructOrError.fromStruct(new HttpFault(faultDelay, faultAbort));
+    }
+  }
+
+  /**
+   * See corresponding Envoy proto message {@link
+   * io.envoyproxy.envoy.extensions.filters.common.fault.v3.FaultDelay}.
+   */
+  static final class FaultDelay {
+    @Nullable
```

Comment:
Instead of using `null` value to indicate that delays are controlled by headers,  an explicit boolean would be better.